site/_posts/2020-11-15-how-to-securely-implement-oauth-rails.md

```diff
@@ -0,0 +1,382 @@
+
+The second attribute is the HMAC signing key that is used to encode the token. Navigate to "Settings" and then "Key Master" from the left navigation bar. 
+Then click on the "View" button that corresponds to the `Default Signing Key`. From here, "click to reveal" the key and copy it to a safe location. 
+We will need this value to successfully decode the token in our application.
+
+{% include _image.liquid src="/assets/img/blogs/fusionauth-example-rails/secret.png" alt="The Edit Tenant page where the issuer field is set." class="img-fluid" figure=false %}
+
+## Build the Rails application
+Let's get started building a very basic Rails application to demonstrate authentication with FusionAuth.
+If you would like to jump straight to checking out the code, see our Github repo [here](https://github.com/FusionAuth/fusionauth-example-rails-oauth).
+
+Create a new Rails application.
+```bash
+rails new rails-fusionauth-app
+```
+
+Our `Gemfile` will use a few additional dependencies so go ahead and add them before we get started.
+```ruby
+gem 'jwt'
+gem 'oauth2'
+```
+
+Install our dependencies
+```bash
+bundle install
+```
+
+### Configuration
+Before we go any further, there are some critical application attributes we will need from our FusionAuth configuration so that our Rails app knows how to successfully communicate with it.
+Specifically:
+* **Client Id**
+* **Client secret**
+* **Authorized redirect URL's**
+* **Logout URL**
+
+You can recall, these can be found under the "OAuth" tab when modifying an application in FusionAuth.
+
+{% include _image.liquid src="/assets/img/blogs/fusionauth-example-rails/edit-application.png" alt="The OAuth tab in the Edit Application page of FusionAuth." class="img-fluid" figure=false %}
+
+Let's set `client_id`, `client_secret`, `idp_url` (identity provider url), and `redirect_uri` values using `development.rb`.
+I chose to set these values here as they will likely change in staging or production environments, but feel free to set them based on your own standards.
+We will also initialize an environment variable that contains our "HMAC secret" that will be used to decode the token received in the authentication process.
+We want to keep this as secure as possible and therefore we will set it at run-time.
+
+```ruby
+  # OAuth configuration
+  config.x.oauth.client_id = "my-client-id"
+  config.x.oauth.client_secret = "my-super-secret-oauth-secret"
+  config.x.oauth.idp_url = "http://localhost:9011/"
+  config.x.oauth.redirect_uri = "http://localhost:3000/oauth2-callback"
+  config.x.oauth.hmac = ENV['HMAC_SECRET']
+```
+
+### Authenticating with FusionAuth
+Before we get into the code, let's take a look at the Authentication Code grant flow as it applies to our Rails app, the user, and FusionAuth:
+
+{% include _image.liquid src="/assets/img/blogs/fusionauth-example-rails/rails-fa-sequence.png" alt="A sequence diagram illustrating the authentication code grant flow." class="img-fluid" figure=false %}
```

Comment:
An image is fine, but you can also make diagrams on the fly if you create plantuml diagrams here: https://github.com/FusionAuth/fusionauth-site/tree/master/site/_diagrams/blogs/update-php-application and include them like this file does: https://raw.githubusercontent.com/FusionAuth/fusionauth-site/master/site/_posts/2020-10-07-updating-crufty-php-application.md

Not necessary to change for this post, but just something to keep in the back of your mind.

Reply:
Sweet, I was wondering if we used something like plantuml.  Definitely noted.